Various Changes

krishmaj opened the pull request:
CRITICAL bug fixes for user permissions ... without this the interface is completely broken
Additional changes to clean up some of the code, move some files around and remove old and unused files.

krishmaj committed:
File cleanup and reorganisation:
- removed 'data-documents' directory, which contained very old prototypes for component travellers and off-DB JSON files, as well as user-specific files that don't belong in the DB code
- removed directory containing 'simple' pug files, which are no longer needed since it is now much easier to prototype and test changes to the main pug files directly ... removed corresponding function in the main app file
- renamed the 'static/pages.js' directory to 'static/pages' ... having a directory name with an extension was really bugging me!  Changed references to the directory in pug files as required
- moved the 'scss' directory into the 'static' directory, so that it is a 'sibling' to the existing 'css' one ... changed scss compiler section of the main app file accordingly
- cleaned up code in 'static/pages/edit_form.js' (included in this commit since that file was also renamed as above)
- cleaned up code in 'pug/dashboard.pug' (included in this commit since that file was already modified as above)

krishmaj committed:
CRITICAL PERMISSIONS CHANGES:
- removed the use of 'forms:view' - this was doing the same thing as 'forms:edit', but in a much more opaque way.  Replaced with 'forms:edit' if appropriate, or removed completely otherwise
- removed the use of 'components:create' - there will not be a need to create a component but not then be able to edit it, so having only 'components:edit' is fine
- removed the use of 'tests:edit' and 'jobs:edit' - tests and jobs can no longer be edited once submitted
- cleaned up the 'inadequate permissions' webpage, and renamed file to reflect it being tied to a user's specific permissions
- removed the use of 'docs:create' - there will not be aneed to create a document but not then be able to edit it, so having only 'docs:edit' is fine

krishmaj committed:
Cleaning up code and display:
- added a visible space between the Formio render and submission button on the type form editing pages ... looks better
- cleaned up code and interface display on user profile page
- cleaned up code on splash page (no changes to interface display)
- small consistency change in user routes code

micchickenburger commented on app/app.js:
```diff
@@ -102,13 +102,13 @@ async function createApp(app) {
       logToConsole: false // If true, will log to logger.error on errors
   }));
-  app.use('/css',express.static('../scss'));
+  app.use('/css',express.static('../static/scss'));
```
I know we talked about this via email.  It doesn’t matter so much where SCSS files are located; however, they are _not_ static resources.  They have to be compiled.

krishmaj replied:
Good point, yes - I didn't consider that.  Guess there isn't really any existing place where the scss files would fit, so I'll revert it back to how it was originally.

micchickenburger commented on app/lib/permissions.js:
```diff
@@ -65,7 +65,7 @@ function checkPermissionOrUserId(required_permission) {
     if(hasPermission(req,required_permission)) return next();
     var p = required_permission;
-     return res.status(400).render('permissionsError.pug',{required_permission:p,required_user:req.params.user_id});
+     return res.status(400).render('user_permissionError.pug',{required_permission:p,required_user:req.params.user_id});
```
Nitpick:  mix between camel and pascal case.  JavaScript world typically standardizes on camel case.

krishmaj replied:
Noted - will deal with that in the future ... there's quite a few other pug files with the same mix.  Personally I like that it makes it clear the broad ("user") vs. specific ("permissionError") purpose of the file, but that's just me I suppose.

micchickenburger commented on app/pug/user_profile.pug:
```diff
@@ -52,12 +50,11 @@ block content
       dd
-        <pre>#{JSON.stringify(user.permissions)}</pre>
-    
+        code #{JSON.stringify(user.permissions)}
```
Semantic nitpick:  `code` is for displaying computer code.  If the idea is to just use a monospace font, use CSS.  If the semantic is preformatted text, `pre` is the correct element to use.  See here:  https://developer.mozilla.org/en-US/docs/Web/HTML/Element/code

krishmaj replied:
Noted, thanks.  Will change it to use CSS and add to the PR.

micchickenburger reviewed (changes requested):
Two things are just nitpicks, but I do think the scss directory choice should be addressed.

krishmaj committed:
Changes from PR Review:
- reverting location of the scss file(s) ... no good location for them other than where they already are I guess
- changing incorrect use of the 'code' HTML tag to CSS-defined monospace
- also noticed that the lack of spacing between permissions when listed makes readability quite poor ... fixed
]

krishmaj committed:
Found a couple of other places where the display should have been using CSS monospace font instead of 'pre' or 'code'

krishmaj committed:
Removing icons and associated pug file and loading code in app file ... usage of icons was removed a while ago, so the icons themselves and associated display page and loading code no longer do anything.

micchickenburger commented on app/static/css/main.css:
```diff
@@ -6,6 +6,10 @@ pre {
   white-space: pre-wrap;
 }
 
```
I think you need a dot in here to specify a class (we can’t invent new HTML elements.)  Also, a better name for the class might just be `.monospace`

krishmaj replied:
Changed in new commit.

micchickenburger commented on app/static/css/main.css:
```diff
@@ -6,6 +6,10 @@ pre {
 }
 
+codelike {
+  font-family: monospace, monospace;
```
Listing monospace twice is redundant

krishmaj replied:
See here: https://stackoverflow.com/questions/38781089/font-family-monospace-monospace .

micchickenburger replied:
Ah, very interesting.  I think you should add a comment with that link to explain that since it’s not obvious and maybe not well-known.

krishmaj replied:
Added.

micchickenburger commented on app/pug/user_profile.pug:
```diff
@@ -52,12 +50,11 @@ block content
       dd
-        <pre>#{JSON.stringify(user.permissions)}</pre>
-    
+        codelike #{JSON.stringify(user.permissions).replace(/,/g, ", ")}
```
As mentioned below we can’t invent new HTML elements.  Adding a dot in front would establish this as a class

krishmaj replied:
Changed in new commit.

krishmaj committed:
Changing to use a new CSS class for monospace styling, and renamed instances accordingly.

krishmaj committed:
Added some commentary on the double monospace CSS declaration.

krishmaj committed:
Continuing cleanup of library files:
- cleaned up and made consistent the remaining components, tests and jobs library functions
- cleaned up and made consistent some of the other libraries - mainly the smaller ones ... others will be dealt with at a later date
- changed 'referenceComponents' to 'relatedComponents' in Components schema, to keep in line with naming elsewhere in the DB code
- removed 'decompose' and 'decompose_ideas' libraries, which are not being used anywhere or by anything

krishmaj committed:
Bug Fix for conflicting routes:
- previously had two routes '/tests/myDrafts' and '/tests/:formId' with identical templates, which was leading to an incorrect routing when deleting user test drafts.  Fixed by changing latter to explicitly direct to '/tests/:formId/list' instead
- changed component and job routes to be equivalent to the above for tests, so as to keep the page addresses consistent across entities
- cleaned up 'autocomplete' routes file
- updated development team message on homepage to give a little more detail

micchickenburger commented on app/lib/constants.js:
```diff
@@ -1,39 +1,43 @@
   SESSION_SECRET: process.env.SESSION_SECRET || crypto.randomBytes(128).toString('hex'),
   M2M_SECRET: process.env.M2M_SECRET || crypto.randomBytes(256).toString('hex'),
-
+  
```
You’ve introduced a lot of rogue whitespace in this file.  My assumption is that you’re not using VS Code yet?  I wrote this file to be compatible with future styling configuration so it really should not be changing at all unless you’ve found a bug or are adding a new feature.

micchickenburger commented on app/lib/email.js:
```diff
@@ -1,28 +1,18 @@
+              subject: "DUNE APA DB Message " + moment().toISOString(),
+              to     : EMAIL.ADMIN_RECIPIENTS,
+              ...arg};
+  
```
Also here you are moving from a standard JavaScript style to an nonstandard one.  Note also that when we implement linting all this work would be automated.

micchickenburger reviewed (changes requested):
I feel compelled to reject the changes in this recent commit since they are stylistic in nature (they don’t fix any bugs or add new features) and are contrary to typical style standards.  Implementing linting can help us automate all of these stylistic and syntactical changes to industry standards.

krishmaj committed:
Reverting my incorrect JavaScript styling changes in library files, except for the 'components', 'tests' and 'jobs' files since they also contain important bug fixes.

krishmaj commented:
As per our email discussion, I've completely reverted all the library files that I made only stylistic changes to, leaving only `Components,js`, `Tests.js` and `Jobs.js` using my non-standard styling, since I don't want to revert my bug fixes in those.  Very sorry about all that ... as I mentioned - I'm having trouble getting VS Code working on my computer (is a problem with the computer, not VS Code), so I haven't switched over to that yet.
